bin/propolis-server/src/lib/vm/state_driver.rs

```diff
@@ -124,17 +128,9 @@ where
     ) -> tokio::sync::watch::Sender<ApiMonitoredState> {
         info!(self.log, "State worker launched");
 
-        // Allow actions to queue up state changes that are applied on the next
-        // loop iteration.
-        let mut next_lifecycle: Option<LifecycleStage> = None;
         loop {
-            let event = self.controller.wait_for_next_event();
-            let outcome = self.handle_event(event, &mut next_lifecycle);
-
-            if let Some(next_lifecycle) = next_lifecycle.take() {
-                self.controller.set_lifecycle_stage(next_lifecycle);
-            }
-
+            let event = self.shared_state.wait_for_next_event();
+            let outcome = self.handle_event(event);
```

Comment:
This pattern is definitely much cleaner than what was here before -- it looks nice!